File: backend/app/models/account.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID, uuid4

from sqlalchemy import (
    DateTime,
    ForeignKey,
    Numeric,
    String,
    UniqueConstraint,
    func,
)
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base


class Account(Base):
    __tablename__ = "accounts"

    __table_args__ = (
        UniqueConstraint(
            "user_id",
            "asset_id",
            "account_type",
            name="uq_user_asset_account_type",
        ),
    )

    id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        primary_key=True,
        default=uuid4,
    )

    user_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    asset_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("assets.id"),
        nullable=False,
        index=True,
    )

    account_type: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
    )

    available_balance: Mapped[Decimal] = mapped_column(
        Numeric(38, 18),
        nullable=False,
        default=Decimal("0"),
    )

    locked_balance: Mapped[Decimal] = mapped_column(
        Numeric(38, 18),
        nullable=False,
        default=Decimal("0"),
    )

    status: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        default="ACTIVE",
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

File: mobile/lib/models/account.dart
class Account {
  const Account({
    required this.id,
    required this.assetId,
    required this.accountType,
    required this.availableBalance,
    required this.lockedBalance,
    required this.totalBalance,
    required this.status,
  });

  final String id;
  final String assetId;
  final String accountType;
  final double availableBalance;
  final double lockedBalance;
  final double totalBalance;
  final String status;

  factory Account.fromJson(Map<String, dynamic> json) => Account(
        id: json['id'].toString(),
        assetId: json['asset_id'].toString(),
        accountType: json['account_type']?.toString() ?? '',
        availableBalance: _number(json['available_balance']),
        lockedBalance: _number(json['locked_balance']),
        totalBalance: _number(json['total_balance']),
        status: json['status']?.toString() ?? '',
      );

  static double _number(dynamic value) => double.tryParse(value?.toString() ?? '') ?? 0;
}
